### PR TITLE
createrepo_c shouldn't silently produce duplicate-NEVRA repos

### LINKED ISSUE #307: createrepo_c does not warn or error when building a repository where the same package is present more than once
My understanding is that "pkgid" and package NEVRA are meant to be unique in a repo.  createrepo_c (the command line tool, at least) ought to provide a warning or error when commanded to build a repo with the same package present twice

This can occur when:

* The user thinks they can just rename the RPM to make it a different package (the package metadata is in the header so obviously this doesn't work)
* The user puts the same RPM in multiple sub-directories

This is frequently seen in the wild even in commonly used community repos (centos7-opstools) and professional repos (prometheus, gitlab, perf-sonar, RHEL6, microsoft azure RHEL additions)

### PR DESCRIPTION
**Pull request by dralley**
Return an error code and print a message when more than one package have
the same NEVRA.

closes #307

### CODE REVIEW

File: src/createrepo_c.c [hunk 1]
```diff
@@ -427,10 +427,10 @@ cr_create_repomd_records_for_groupfile_metadata(const cr_Metadatum *group_metada
 }
 
 /** Creates list of cr_RepomdRecords from list
- *  of additional metadata (cr_Metadatum) 
+ *  of additional metadata (cr_Metadatum)
  *
  * @param additional_metadata       List of cr_Metadatum
- * @param repomd_checksum_type
```
Comment (dralley) [inline src/createrepo_c.c:433]: Sorry about this, my editor did it, I can get rid of these changes if you want.

File: src/createrepo_c.c [hunk 2]
```diff
@@ -1306,9 +1308,34 @@ main(int argc, char **argv)
...
+    GHashTableIter iter;
+    gpointer key, value;
+
+    g_hash_table_iter_init(&iter, user_data.nevra_table);
+    while (g_hash_table_iter_next(&iter, &key, &value))
+    {
+        gchar *nevra = (gchar *) key;
+        GArray *locations = (GArray *) value;
+        if (locations->len > 1) {
+            g_warning("Package '%s' has duplicate metadata entries, only one should exist", nevra);
```
Comment (dralley) [inline src/createrepo_c.c:1320]: `g_error()` makes it terminate immediately, which I would be somewhat fine with, except that it prevents logging information on how to fix the problem.

Comment (dralley) [reply src/createrepo_c.c:1320]:
```
(.env) [dalley@thinkpad test]$ createrepo_c .
Directory walk started
Directory walk done - 3 packages
Temporary output repo path: ./.repodata/
Preparing sqlite DBs
Pool started (with 5 workers)
Warning: Package 'wget-0:1.21.3-1.fc35.x86_64' has duplicate metadata entries, only one should exist
Warning:     Sourced from location: 'bar/wget-1.21.3-1.fc35.x86_64.rpm'
Warning:     Sourced from location: 'foo/wget-1.21.3-1.fc35.x86_64.rpm'
Warning:     Sourced from location: 'wget-1.21.3-1.fc35.x86_64.rpm'

```

File: src/createrepo_c.c (hunk 1 above)
Comment (kontura) [reply src/createrepo_c.c:433]: I think its not a problem here.

File: src/createrepo_c.c [hunk 3]
```diff
@@ -1286,6 +1287,7 @@ main(int argc, char **argv)
     user_data.had_errors        = 0;
     user_data.output_pkg_list   = output_pkg_list;
 
+    g_mutex_init(&(user_data.mutex_nevra_table));
```
Comment (kontura) [inline src/createrepo_c.c:1290]: We have to also `g_mutex_clear()` the `user_data.mutex_nevra_table`, its around [here](https://github.com/rpm-software-management/createrepo_c/blob/master/src/createrepo_c.c#L1452-L1455).

**Comment by Conan-Kudo** [comment]
This will break COPR. cc: @praiskup

**Review by kontura — CHANGES_REQUESTED** [review]
I have also created CI test update: https://github.com/rpm-software-management/ci-dnf-stack/pull/1129

File: src/createrepo_c.c (hunk 3 above)
Comment (dralley) [reply src/createrepo_c.c:1290]: @kontura Done

**Comment by dralley** [comment]
@kontura Left a comment here https://github.com/rpm-software-management/ci-dnf-stack/pull/1129/files#r903764740=

**Comment by dralley** [comment]
Would this be unreasonable to backport to, say, EL8 and EL9?  (or not backport, but rather, rebase to the latest available)

File: src/createrepo_c.c [hunk 4]
```diff
@@ -1306,9 +1308,51 @@ main(int argc, char **argv)
...
+                if (a_len > b_len) {
+                    return 1;
+                } else if (b_len > a_len) {
+                    return -1;
+                } else {
+                    return 0;
+                }
+            }
+
+            g_array_sort(locations, strsort);
```
Comment (dralley) [inline src/createrepo_c.c:1337]:
```
(.env) [dalley@thinkpad test]$ createrepo_c .
Directory walk started
Directory walk done - 5 packages
Temporary output repo path: ./.repodata/
Preparing sqlite DBs
Pool started (with 5 workers)
Warning: Package 'htop-0:3.2.1-1.fc35.x86_64' has duplicate metadata entries, only one should exist
Warning:     Sourced from location: 'htop-3.2.1-1.fc35.x86_64.rpm'
Warning:     Sourced from location: 'htop-3.2.1-1.fc35_rename.x86_64.rpm'
Warning: Package 'wget-0:1.21.3-1.fc35.x86_64' has duplicate metadata entries, only one should exist
Warning:     Sourced from location: 'wget-1.21.3-1.fc35.x86_64.rpm'
Warning:     Sourced from location: 'foo/wget-1.21.3-1.fc35.x86_64.rpm'
Warning:     Sourced from location: 'bar/wget-1.21.3-1.fc35.x86_64.rpm'
Pool finished with errors
```

**Comment by j-mracek** [comment]
The change looks logical but it will brake Copr and other workflows that previously worked (main reason why I marked it as blocked). The change does not describe explicitly why it is required but it is essential. I can only guess that it will provide a performance improvement but it is only my guess.

**Comment by dralley** [comment]
@Conan-Kudo @j-mracek What if we leave the warnings but don't set the error code.

**Comment by j-mracek** [comment]
I think that a warning will not brake anything and it contains interesting piece of information therefore I am fine with a warning. If there will be a complain we can easily turn the warning into info message.

File: src/createrepo_c.c [hunk 5]
```diff
@@ -427,10 +427,10 @@ cr_create_repomd_records_for_groupfile_metadata(const cr_Metadatum *group_metada
 }
 
 /** Creates list of cr_RepomdRecords from list
- *  of additional metadata (cr_Metadatum) 
+ *  of additional metadata (cr_Metadatum)
  *
  * @param additional_metadata       List of cr_Metadatum
- * @param repomd_checksum_type      
+ * @param repomd_checksum_type
```
Comment (praiskup) [inline src/createrepo_c.c:433]: Please avoid unnecessary white space changes.

**Comment by praiskup** [comment]
Thank you for the work here, and for pinging me.   Yes, with non-zero exit status this would start failing in Copr for many repositories, unfortunately.  Warning would be completely OK, and "opt-in" non-zero exit status as well.  If we did this "opt-out", like "--ignore-duplicate-nvrs", we would have to coordinate with an appropriate fix in Copr upstream.

File: src/createrepo_c.c (hunk 4 above)
Comment (praiskup) [reply src/createrepo_c.c:1339]: Just a nit, do we want to sort reverse-alphabetically?

File: src/dumper_thread.c [hunk 6]
```diff
@@ -550,6 +550,19 @@ cr_dumper_thread(gpointer data, gpointer user_data)
...
+    g_mutex_lock(&(udata->mutex_nevra_table));
+    gchar *nevra = cr_package_nevra(pkg);
+    GArray *pkg_locations = g_hash_table_lookup(udata->nevra_table, nevra);
+    if (!pkg_locations) {
+        pkg_locations = g_array_new(FALSE, TRUE, sizeof(gchar *));
+        g_hash_table_insert(udata->nevra_table, nevra, pkg_locations);
+    }
+    gchar *location = g_strdup(pkg->location_href);
+    g_array_append_val(pkg_locations, location);
+    g_mutex_unlock(&(udata->mutex_nevra_table));
```
Comment (praiskup) [inline src/dumper_thread.c:564]: It looks pretty innocent, (O(logN) I hope.  But actually we might need to opt out even the calculating of this hash.  In Copr with the [giant repositories like this](https://download.copr.fedorainfracloud.org/results/@rubygems/rubygems/fedora-rawhide-x86_64/repodata/), we use ` --no-stat --update --recycle-pkglist` to speedup metadata updates, which in turn means that createrepo_c isn't doing fancy with the `pkg` structure.  Now we use it to calculate the hash.

This deserves some benchmark IMO.

Comment (dralley) [reply src/dumper_thread.c:564]: I'm running this command currently with the nonmodified `createrepo_c` and it is taking upwards of 10 minutes, is that expected?

Comment (dralley) [reply src/dumper_thread.c:564]: Yeah, so, this is why.  85% of the runtime is spent inside of `g_queue_insert_sorted()`  That is crazy.  Given this, I'm not sure measuring any additional impact of my code is a great use of time. I don't expect it to even register, given that producing and compressing multiple gigabytes of metadata just barely registers as a tiny blip way over on the right-hand side of the chart.

flamegraph

**Comment by dralley** [comment]
I have removed the error status and fixed the accidental formatting changes.

File: src/dumper_thread.c (hunk 6 above)
Comment (dralley) [reply src/dumper_thread.c:564]: Filed as https://github.com/rpm-software-management/createrepo_c/issues/323

Comment (praiskup) [reply src/dumper_thread.c:564]: Nice analysis, @dralley!

**Comment by praiskup** [comment]
In Copr, we are unable to (or we don't want to) control how our users take care of NVRs. Maybe I'm missing something, do you mean the opt-out/opt-in option?  Feel free to fill the [issue here](https://pagure.io/copr/copr/issues).

File: src/createrepo_c.c [hunk 7]
```diff
@@ -1306,9 +1308,51 @@ main(int argc, char **argv)
...
+    gpointer key, value;
+
+    g_hash_table_iter_init(&iter, user_data.nevra_table);
+    while (g_hash_table_iter_next(&iter, &key, &value))
+    {
+        gchar *nevra = (gchar *) key;
+        GArray *locations = (GArray *) value;
+        if (locations->len > 1) {
+            g_warning("Package '%s' has duplicate metadata entries, only one should exist", nevra);
+            int strsort(const gpointer a, const gpointer b) {
```
[the remark this answers is not in the document]
Comment (kontura) [reply src/createrepo_c.c:1323]: You are correct, for example it doesn't compile with `clang`.
Please define it outside as a `static` function.

**Comment by praiskup** [comment]
I rather meant that Copr allows the users to build whatever NVR they want, even multiple times, without additional constraints like "fail the build when the NVR already exists".  Copr does the build, places the results into an appropriate directory, and runs createrepo.  If createrepo fails, the build fails as well.

**Comment by praiskup** [comment]
*Technically* it is possible (both the build ID and NVR is known and available), but automatic removals are not desired.  All builds are preserved for [at least 14 days](https://docs.pagure.org/copr.copr/user_documentation.html#how-long-do-you-keep-the-builds) - during this period it is up to the project maintainer to remove manually what they prefer.

**Review by kontura — APPROVED** [review]

**Comment by dralley** [comment]
@praiskup Still, this is what confuses me - the shadowed releases would be uninstallable, I don't even know that you could "downgrade" to them as that might require a different NEVRA (not sure about this).

Thus whether they're kept or not is not a very consequential decision, and the choice to keep them might be a false choice as they're not really available in a true sense, unless you go hunt the packages down manually.

**Comment by dralley** [comment]
Sidenote, once the new version makes it to the repos it would be neat to see if the performance fix makes a meaningful difference to the overall COPR performance metrics, if that stuff gets measured.

**Comment by kontura** [comment]
> @praiskup Still, this is what confuses me - the shadowed releases would be uninstallable, I don't even know that you could "downgrade" to them as that might require a different NEVRA (not sure about this).

You can't, at least not with `dnf downgrade` command. I tried it now and it seems that the only way to get the "other" rpm with dnf would be `reinstall` with path directly to it which bypasses the repo anyway. 

> 
> Thus whether they're kept or not is not a very consequential decision, and the choice to keep them might be a false choice as they're not really available in a true sense, unless you go hunt the packages down manually.

I am not sure if this is what you mean and I am just repeating it but perhaps the older builds could be still kept in COPR but they would be excluded from the repo. From user perspective almost nothing would change (reposync wouldn't download the other pkgs anymore?).

**Comment by dralley** [comment]
Right, and that feels like a viable option

>I am not sure if this is what you mean and I am just repeating it but perhaps the older builds could be still kept in COPR but they would be excluded from the repo.

Yup that's exactly what I meant.

**Comment by praiskup** [comment]
Yes, user can manually decide what build in particular (from the duplicated builds) is to be deleted.  The delete action then triggers a build removal, and an automatic createrepo run.  So while the RPMs are not installable all the time they exist, it is just a non-problem situation, and it can become installable again.  This is just a result of some user/scripting error, easily fixable by the user.

Ignoring the directories that provide duplicate RPMs is technically possible, but potentially a very expensive feature (both for computational power and for the eventual code contribution).  And it would bring a zero gain to anyone.

Nb. this would be a serious backward incompatibility anyway, this could twist the createrepo users (not just Copr) into an inconvenient situation.  If the exit status is ever to be switched, we should have some opt-out option to absorb the change.

**Pull request merged**